cpu/stm32_common/periph/uart.c

```diff
@@ -44,19 +46,16 @@ static inline USART_TypeDef *dev(uart_t uart)
     return uart_config[uart].dev;
 }
 
-int uart_init(uart_t uart, uint32_t baudrate, uart_rx_cb_t rx_cb, void *arg)
-{
-    uint16_t mantissa;
-    uint8_t fraction;
-    uint32_t clk;
-
-    assert(uart < UART_NUMOF);
-
-    /* save ISR context */
-    isr_ctx[uart].rx_cb = rx_cb;
-    isr_ctx[uart].arg   = arg;
+static inline void uart_init_usart(uart_t uart, uint32_t baudrate);
+#if defined(CPU_FAM_STM32L0) || defined(CPU_FAM_STM32L4)
+#ifdef MODULE_PERIPH_LPUART
+static inline void uart_init_lpuart(uart_t uart, uint32_t baudrate);
```

Comment:
Any reason for this to be inline?

Reply:
Note that I took this from other piece of code in RIOT (widely used apparently), see stm32 i2c periph driver for example. I guess this is done for optimisation purpose: in theory the compiler replaces the call to an inline function by the actual code of the function. For init function, this is maybe not worth.
If I remove the `inline`, there's no difference in term of code size.

Reply:
`inline` only gives a hint to the compiler to try to optimise the function, it might be ignored here. However, AFAIK, `static inline` should only be used on headers and not C files.

Reply:
> However, AFAIK, static inline should only be used on headers and not C files.

Can you provide a reference for this? Because if it's the case, there's a lot to change in RIOT's code ;)

```
$ git grep "static inline" *.c
525
```

Reply:
I was wrong saying it "should". It is possible but I don't see here the benefits of doing it, since the functions are maybe too big to be inlined (probably the `inline` is ignored by the compiler), as you stated.

For the `inline` doc I was reading this: https://www.greenend.org.uk/rjk/tech/inline.html 

> For init function, this is maybe not worth.
If I remove the inline, there's no difference in term of code size.

I think it would be more correct if you do so.

Reply:
I disagree, using static inline is like having a macro. It is perfectly fine as it is now.

Reply:
I think we need a third advice here, since I'm not 100% convinced about using the `static inline` at this level.

Any opinions @cladmi @kaspar030 @miri64 @vincent-d ?

Reply:
I think the compiler would treat them equally anyway.

Reply:
> I think the compiler would treat them equally anyway.

I agree and this change request is completely pointless. Can we move forward one day and merge this PR ?

Reply:
> I think the compiler would treat them equally anyway.

Yes, but the discussion here is about readability and code style. I won't say this is violating something, just want to be sure that the reasons why the function is declared that way actually exist, and is not just a "it's done this way somewhere else" or "it's perfectly fine".

The change is actually unrelated to the purpose of this PR. If you want to change it to `static inline`, please open another PR to do so. However, for such a PR I won't have the time to review it.

Reply:
As the compiler is free to ignore "inline" and will inline anyways, inline-wise (code-size wise) there's no difference.

A static function that's not "inline" will generate a warning, though, if unused. Thus in headers, functions need to be "static inline" as the function might not be used by every includer.

In ```*.c``` files, without "inline" there'd also be a warning if unused, which might be desirable. Other than that, I personally use "static inline" like @aabadie mentioned: instead of a macro. To me it's a message to the code reader: "this function is super small and I expect it to be inlined".

TL;DR the inline is fine here, the change request should at most be postponed into a general issue.

Reply:
Thanks for the explanation @kaspar030! Would you mind to ack the code style? I can ACK the rest and merge it if other people don't have other requests.